geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/DeltaSet.java

```diff
@@ -0,0 +1,261 @@
+      throws IOException, InvalidDeltaException {
+    boolean deltaAdds = DataSerializer.readBoolean(in);
+    try {
+      ArrayList<ByteArrayWrapper> deltas = DataSerializer.readArrayList(in);
+      if (deltas != null) {
+        if (deltaAdds) {
+          members.addAll(deltas);
+        } else {
+          members.removeAll(deltas);
+        }
+      }
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // DATA SERIALIZABLE
+
+  @Override
+  public void toData(DataOutput out) throws IOException {
+    DataSerializer.writeHashSet(members, out);
+  }
+
+  @Override
+  public void fromData(DataInput in) throws IOException, ClassNotFoundException {
+    members = DataSerializer.readHashSet(in);
+  }
+
+
+  /**
+   * @param membersToAdd members to add to this set; NOTE this list may by
+   *        modified by this call
+   * @param region the region this instance is stored in
+   * @param key the name of the set to add to
+   * @return the number of members actually added
+   * @throws RetryDueToConcurrentModification if a concurrent modification is detected
+   */
+  private synchronized long saddInstance(ArrayList<ByteArrayWrapper> membersToAdd,
+      Region<ByteArrayWrapper, DeltaSet> region,
+      ByteArrayWrapper key) {
+    if (region.get(key) != this) {
+      throw new RetryDueToConcurrentModification();
+    }
+    membersToAdd.removeIf(memberToAdd -> !members.add(memberToAdd));
+    int membersAdded = membersToAdd.size();
+    if (membersAdded != 0) {
+      deltasAreAdds = true;
+      deltas = membersToAdd;
+      try {
+        region.put(key, this);
+      } finally {
+        deltas = null;
+      }
+    }
+    return membersAdded;
+  }
+
+  /**
+   * @param membersToRemove members to remove from this set; NOTE this list may by
+   *        modified by this call
+   * @param region the region this instance is stored in
+   * @param key the name of the set to remove from
+   * @return the number of members actually removed
+   * @throws RetryDueToConcurrentModification if a concurrent modification is detected
+   */
+  private synchronized long sremInstance(ArrayList<ByteArrayWrapper> membersToRemove,
+      Region<ByteArrayWrapper, DeltaSet> region,
+      ByteArrayWrapper key) {
+    if (region.get(key) != this) {
+      throw new RetryDueToConcurrentModification();
+    }
+    membersToRemove.removeIf(memberToRemove -> !members.remove(memberToRemove));
+    int membersRemoved = membersToRemove.size();
+    if (membersRemoved != 0) {
+      deltasAreAdds = false;
+      deltas = membersToRemove;
+      try {
+        region.put(key, this);
+      } finally {
+        deltas = null;
+      }
+    }
+    return membersRemoved;
+  }
+
+  /**
+   * This exception is thrown if a modification fails because some other
+   * thread changed what is stored in the region.
+   */
+  static class RetryDueToConcurrentModification extends RuntimeException {
```

Comment:
This exception name tells the User what to do, i.e. "retry", instead of just stating what happened and letting the user decide how to handle it.

Maybe just throwing a ConcurrentModificationException, or something like RegionUnexpectedModificationException

Reply:
Keep in mind this exception is for our internal implementation; not the user. It is possible we could refactor this code so that we don't need to throw on exception to force a retry.